Split off upgrade and plugins bats tests

rjernst opened the pull request:
The bats tests currently require many additional artifacts to be built.
In addition to the current distributions, they need all the plugins to
be installed, as well as a randomly chosen bwc distribution. This commit
splits these two cases into their own bats task, so the dependencies do
not slow down other tasks like distroTests which do not need them.

rjernst committed:
Split off upgrade and plugins bats tests

The bats tests currently require many additional artifacts to be built.
In addition to the current distributions, they need all the plugins to
be installed, as well as a randomly chosen bwc distribution. This commit
splits these two cases into their own bats task, so the dependencies do
not slow down other tasks like distroTests which do not need them.

elasticmachine commented:
Pinging @elastic/es-core-infra

rjernst commented:
@elasticmachine run elasticsearch-ci/1 and run elasticsearch-ci/2

mark-vieira commented on buildSrc/src/main/groovy/org/elasticsearch/gradle/test/DistroTestPlugin.java:
```diff
@@ -66,13 +67,21 @@
     // all distributions used by distro tests. this is temporary until tests are per distribution
-    private static final String PACKAGING_DISTRIBUTION = "packaging";
+    private static final String PACKAGING_CONFIGURATION = "packaging";
+    private static final String UPGRADE_CONFIGURATION = "packaging_upgrade";
```
Can we use camel-case for these like we do for most everything else?

mark-vieira commented on buildSrc/src/main/groovy/org/elasticsearch/gradle/test/DistroTestPlugin.java:
```diff
@@ -164,9 +177,30 @@ private void configureCopyPackagingTask(Project project) {
+
+    private void configureCopyUpgradeTask(Project project) {
+        this.upgradeDir = project.getParent().getLayout().getBuildDirectory().dir("packaging/upgrade");
+        // temporary, until we have tasks per distribution
```
Does this mean we intend to ditch this assumption that all distro test projects will be under a common parent? Why not just use a similar pattern as we do elsewhere and add this to the root project so there's no risk of there ever being more than one?

rjernst replied:
This is already in the root project. All of these copy tasks, and their associated target directories, are part of qa/vagrant.

mark-vieira commented on buildSrc/src/main/groovy/org/elasticsearch/gradle/test/DistroTestPlugin.java:
```diff
@@ -164,9 +177,30 @@ private void configureCopyPackagingTask(Project project) {
+        this.copyUpgradeArchives = Boilerplate.maybeRegister(project.getParent().getTasks(), COPY_UPGRADE_TASK, Copy.class,
+            t -> {
+                t.into(upgradeDir);
+                t.from(project.getConfigurations().getByName(UPGRADE_CONFIGURATION));
```
Something doesn't seem quite right here. We create this configuration per-project, but we only every create a single one of these tasks. Essentially, the first distro test project to be configured "wins" here. Is that the intention?

rjernst replied:
Sorry this is a bug. The intention was to have the configuration on the parent (qa/vagrant).

rjernst commented:
@elasticmachine run elasticsearch-ci/1

mark-vieira commented on buildSrc/src/main/groovy/org/elasticsearch/gradle/test/DistroTestPlugin.java:
```diff
@@ -178,21 +212,30 @@ private void configureCopyPackagingTask(Project project) {
-                inputs.property("version", VersionProperties.getElasticsearch());
                 inputs.property("upgrade_from_version", upgradeFromVersion);
                 // TODO: this is serializable, need to think how to represent this as an input
                 //inputs.property("bwc_versions", bwcVersions);
```
We can track `BwcVersions` by breaking it up into it's constituent parts. Alternatively, we could track the input to `BwcVersions` _itself_ which is simply `Version.java`

rjernst replied:
I'll do that as a followup since this already existed from my previous PR.

mark-vieira commented on buildSrc/src/main/groovy/org/elasticsearch/gradle/test/DistroTestPlugin.java:
```diff
@@ -66,13 +67,21 @@
 
     private static Version upgradeVersion;
     private Provider<Directory> archivesDir;
+    private Provider<Directory> upgradeDir;
```
I think we should initialize these providers early on in the `apply()` method. By doing it later we can very easily fuck things up by reordering when we call these various `configure*()` methods.

mark-vieira commented on buildSrc/src/main/groovy/org/elasticsearch/gradle/test/DistroTestPlugin.java:
```diff
@@ -90,12 +99,16 @@ public void apply(Project project) {
+        configureBatsTest(project, "oss", null, null, copyPackagingArchives, systemJdk, gradleJdk);
+        configureBatsTest(project, "plugins", pluginsDir.get(), null, copyPackagingArchives, copyPlugins, systemJdk, gradleJdk);
+        configureBatsTest(project, "upgrade", null, upgradeDir.get(), copyPackagingArchives, copyUpgradeArchives, systemJdk, gradleJdk);
+        configureBatsTest(project, "default", null, null,copyPackagingArchives, systemJdk, gradleJdk);
```
nit: we are missing a space between arguments here.

mark-vieira commented on buildSrc/src/main/groovy/org/elasticsearch/gradle/test/DistroTestPlugin.java:
```diff
@@ -247,51 +292,60 @@ private void configureBatsTest(Project project, String type) {
         // temporary until distro tests have one test per distro
-        Configuration packagingConfig = project.getConfigurations().create(PACKAGING_DISTRIBUTION);
-        List<Configuration> distroConfigs = distributions.stream().map(ElasticsearchDistribution::getConfiguration)
+        Configuration packagingConfig = project.getConfigurations().create(PACKAGING_CONFIGURATION);
```
It might be a good idea to use detached configurations here. Realistically, these configurations are purely for internal use and never intended for users to configure so we might as well hide them from folks entirely.

rjernst replied:
I've left this for now because these are strictly temporarily configurations. Once we have distroTest tasks per distribution, that will depend on the distribution directly and we won't have a giant copy task anymore.

mark-vieira commented on buildSrc/src/main/java/org/elasticsearch/gradle/test/BatsTestTask.java:
```diff
@@ -85,6 +108,12 @@ public void runBats() {
             spec.environment(System.getenv());
             spec.environment("BATS_TESTS", testsDir.getAsFile().toString());
             spec.environment("BATS_UTILS", utilsDir.getAsFile().toString());
+            if (pluginsDir != null) {
```
Is there any harm in _always_ passing these directories and environment variables? Wouldn't the tests for which these are non-applicable just ignore them?

rjernst replied:
I tried this, but Gradle complained that the directory did not exist, since in several cases the task that sets it up won't have been run.

mark-vieira commented on qa/vagrant/build.gradle:
```diff
@@ -102,30 +102,26 @@ subprojects { Project platformProject ->
-copyPackagingArchives {
-  from configurations.allPlugins
+copyPlugins {
+  inputs.property("plugins", pluginNames)
```
Could we simply _infer_ all this nonsense from all the dependencies added to the plugins configuration?

rjernst replied:
I removed the inputs, since that is inferred by the configuration. But we need to still write the file, since that is currently required by the bats upgrade tests.

mark-vieira reviewed (approved):
Handful of comments, otherwise LGTM.

rjernst committed:
address feedback

rjernst committed:
only maybe create configurations

rjernst committed:
Make distro test plugin apply to the top level project

The distro test plugin was originally designed to be applied within each
subproject, per operating system we run in a VM with vagrant. However,
for efficiency, and also ease of having a single task to run in CI when
launching within individual OS VMs, having the "destructive" tasks in a
single place is more convenient. This commit reworks the distro test
plugin to be applied to the qa/vagrant project, which now creates only
the wrapper tasks in each of the subprojects for each vagrant VM.

elasticmachine committed:
Merge branch 'master' into distro_tests20

rjernst committed:
adjust forbiddenapis test signatures now instead of main

rjernst committed:
Merge branch 'master' into distro_tests18

rjernst committed:
Merge branch 'master' into distro_tests20

rjernst committed:
Merge branch 'distro_tests20' into distro_tests18

rjernst committed:
convert to new style with tasks on parent project

rjernst committed:
fix incorrect ref to parent project

rjernst committed:
Merge branch 'master' into distro_tests20

rjernst committed:
address feedback

rjernst committed:
fix checkstyle

rjernst committed:
Merge branch 'master' into distro_tests20

rjernst committed:
Merge branch 'distro_tests20' into distro_tests18

rjernst committed:
fix forgotten calls to get()

rjernst committed:
Merge branch 'master' into distro_tests18
